engineapi/backup_backing_image.go

```diff
@@ -0,0 +1,260 @@
+	if err != nil {
+		if types.ErrorIsNotFound(err) {
+			return nil, nil
+		}
+		return nil, errors.Wrapf(err, "failed to get backup backing image config %s", backupBackingImageURL)
+	}
+	return parseBackupBackingImageConfig(output)
+}
+
+// BackupBackingImageNameList returns a list of backup backing image names
+func (btc *BackupTargetClient) BackupBackingImageNameList() ([]string, error) {
+	output, err := btc.ExecuteEngineBinary("backup", "ls-backing-image", btc.URL)
+	if err != nil {
+		if types.ErrorIsNotFound(err) {
+			return nil, nil
+		}
+		return nil, errors.Wrapf(err, "failed to list backup backing image names")
+	}
+	return parseBackupBackingImageNamesList(output)
+}
+
+// BackupBackingImageDelete deletes the backup volume from the remote backup target
+func (btc *BackupTargetClient) BackupBackingImageDelete(backupURL string) error {
+	_, err := btc.ExecuteEngineBinaryWithoutTimeout("backup", "rm-backing-image", backupURL)
+	if err != nil {
+		if types.ErrorIsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "failed to delete backup backing image")
+	}
+	return nil
+}
+
+type BackupBackingImageMonitor struct {
+	logger logrus.FieldLogger
+
+	backupBackingImageName string
+	client                 *BackingImageManagerClient
+
+	backupBackingImageStatus longhorn.BackupBackingImageStatus
+	backupStatusLock         sync.RWMutex
+
+	syncCallback func(key string)
+	callbackKey  string
+
+	ctx  context.Context
+	quit context.CancelFunc
+}
+
+func NewBackupBackingImageMonitor(logger logrus.FieldLogger, ds *datastore.DataStore, bbi *longhorn.BackupBackingImage, backingImage *longhorn.BackingImage, backupTargetClient *BackupTargetClient,
+	compressionMethod longhorn.BackupCompressionMethod, concurrentLimit int, bimClient *BackingImageManagerClient, syncCallback func(key string)) (*BackupBackingImageMonitor, error) {
+	ctx, quit := context.WithCancel(context.Background())
+	m := &BackupBackingImageMonitor{
+		logger: logger.WithFields(logrus.Fields{"backupBackingImage": bbi.Name}),
+
+		backupBackingImageName: bbi.Name,
+		client:                 bimClient,
+
+		backupStatusLock: sync.RWMutex{},
+
+		syncCallback: syncCallback,
+		callbackKey:  bbi.Namespace + "/" + bbi.Name,
+
+		ctx:  ctx,
+		quit: quit,
+	}
+
+	// Call backing image manager API snapshot backup
+	if bbi.Status.State == longhorn.BackupStateNew {
+		err := m.client.BackupCreate(bbi.Name, backingImage.Status.UUID, bbi.Status.Checksum,
+			backupTargetClient.URL, bbi.Spec.Labels, backupTargetClient.Credential, string(compressionMethod), concurrentLimit)
+		if err != nil {
+			if !strings.Contains(err.Error(), "DeadlineExceeded") {
+				m.logger.WithError(err).Warn("failed to take backing image backup")
+				return nil, err
+			}
+
+			m.logger.WithError(err).Warnf("backup backing image request timeout")
+			bbi.Status.State = longhorn.BackupStatePending
+		}
+
+		m.backupBackingImageStatus = bbi.Status
+	}
+
+	go m.monitorBackupBackingImage()
```

Comment:
If the monitor has been deleted from the backup backing image controller, will this goroutine be deleted immediately or until the timeout?

We should inject the ctx to let the goroutine be deleted immediately.

Reply:
Sinc we have other monitor mechanisms, are they having this issue?

Reply:
Yes, the go routine `m.monitorBackupBackingImage()`
m->monitor has context and a quit() function in it
```
	m := &BackupBackingImageMonitor{
		logger: logger.WithFields(logrus.Fields{"backupBackingImage": bbi.Name}),

		backupBackingImageName: bbi.Name,
		client:                 bimClient,

		backupStatusLock: sync.RWMutex{},

		syncCallback: syncCallback,
		callbackKey:  bbi.Namespace + "/" + bbi.Name,

		ctx:  ctx,
		quit: quit,
	}
```

inside the goroutine, it will return when the `m.ctx` is done
- https://github.com/longhorn/longhorn-manager/pull/2416/files#diff-241fdec810f0d17f4f8d4857a52ea99eb1cf318e00992cd0a530c64738ca7676R148
- https://github.com/longhorn/longhorn-manager/pull/2416/files#diff-241fdec810f0d17f4f8d4857a52ea99eb1cf318e00992cd0a530c64738ca7676R207

Reply:
yeah, it will be called by `quite`.